Sources/ForTag.swift

```diff
@@ -124,25 +131,115 @@ class ForNode : NodeType {
 
     if !values.isEmpty {
       let count = values.count
+      var result = ""
+
+      var currentLoopContext = context["forloop"] as? [String: Any]
 
-      return try values.enumerated().map { index, item in
-        let forContext: [String: Any] = [
+      for (index, item) in values.enumerated() {
+        var forContext: [String: Any] = [
           "first": index == 0,
           "last": index == (count - 1),
           "counter": index + 1,
-          "counter0": index,
-        ]
+          "counter0": index
+          ]
+        if let label = label {
+          forContext[label] = forContext
+          forContext["label"] = label
+        }
+        if let label = currentLoopContext?["label"] as? String,
+          let labeledContext = currentLoopContext?[label] {
+          forContext[label] = labeledContext
+        }
 
-        return try context.push(dictionary: ["forloop": forContext]) {
-          return try push(value: item, context: context) {
+        var shouldBreak: Bool = false
+        result += try context.push(dictionary: ["forloop": forContext]) {
+          let result = try push(value: item, context: context) {
             try renderNodes(nodes, context)
           }
+          shouldBreak = context[LoopTerminationNode.breakContextKey] != nil
```

Comment:
Maybe move the shouldBreak calculation logic out to a small method to keep this part a bit cleaner.